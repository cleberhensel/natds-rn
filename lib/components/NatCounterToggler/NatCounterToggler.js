import React, { useState } from 'react';
import { Platform, TouchableNativeFeedback } from 'react-native';
import PropTypes from 'prop-types';

import { colors } from '../../assets/theme/colors/colors';
import {
  CounterView,
  ContentViewDefault,
  ContentViewLeft,
  ContentViewRight,
  CustomTouchableHighlight,
  CustomTextOperators,
} from './NatCounterToggler.styles';

export const NatCounterToggler = ({
  onIncrease, onDecrease, counter, color, width, height, min, max,
}) => {
  const [decreaseDisabled, setDecreaseDisabled] = useState(true);
  const [increaseDisabled, setIncreaseDisabled] = useState(false);
  const [decreaseColor, setDecreaseColor] = useState(colors.bw.grayLight);
  const [increaseColor, setIncreaseColor] = useState(colors.primary.white);

  const actionHandler = (action) => {
    if (action === 'increase') {
      onIncrease();
      if (counter === min) {
        setDecreaseColor(colors.primary.white);
        setDecreaseDisabled(false);
      }
      if ((counter + 1) === max) {
        setIncreaseColor(colors.bw.grayLight);
        setIncreaseDisabled(true);
      }
    } else {
      onDecrease();
      if (counter === max) {
        setIncreaseColor(colors.primary.white);
        setIncreaseDisabled(false);
      }
      if ((counter - 1) === min) {
        setDecreaseColor(colors.bw.grayLight);
        setDecreaseDisabled(true);
      }
    }
  };

  return Platform.OS === 'android' ? (
    <CounterView color={color} width={width} height={height}>
      <TouchableNativeFeedback disabled={decreaseDisabled} onPress={() => actionHandler('decrease')}>
        <ContentViewLeft color={decreaseColor}>
          <CustomTextOperators>-</CustomTextOperators>
        </ContentViewLeft>
      </TouchableNativeFeedback>

      <ContentViewDefault>
        <CustomTextOperators>{counter}</CustomTextOperators>
      </ContentViewDefault>

      <TouchableNativeFeedback disabled={increaseDisabled} onPress={() => actionHandler('increase')}>
        <ContentViewRight color={increaseColor}>
          <CustomTextOperators>+</CustomTextOperators>
        </ContentViewRight>
      </TouchableNativeFeedback>
    </CounterView>
  ) : (
    <CounterView color={color}>
      <CustomTouchableHighlight disabled={decreaseDisabled} onPress={() => actionHandler('decrease')}>
        <ContentViewLeft color={decreaseColor}>
          <CustomTextOperators>-</CustomTextOperators>
        </ContentViewLeft>
      </CustomTouchableHighlight>

      <ContentViewDefault>
        <CustomTextOperators>{counter}</CustomTextOperators>
      </ContentViewDefault>

      <CustomTouchableHighlight disabled={increaseDisabled} onPress={() => actionHandler('increase')}>
        <ContentViewRight color={increaseColor}>
          <CustomTextOperators>+</CustomTextOperators>
        </ContentViewRight>
      </CustomTouchableHighlight>
    </CounterView>
  );
};

NatCounterToggler.propTypes = {
  onIncrease: PropTypes.func.isRequired,
  onDecrease: PropTypes.func.isRequired,
  counter: PropTypes.string.isRequired,
  color: PropTypes.string.isRequired,
  width: PropTypes.string,
  height: PropTypes.string,
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
};
NatCounterToggler.defaultProps = {
  width: null,
  height: null,
};
