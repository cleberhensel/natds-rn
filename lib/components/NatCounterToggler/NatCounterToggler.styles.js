import styled from 'styled-components/native';
import { colors } from '../../assets/theme/colors/colors';

const CommonContentView = `
    flex: 1;
    justify-content: center;
    align-items: center;
`;

export const CounterView = styled.View`
    flex-direction: row;
    align-items: stretch;
    justify-content: space-between;
    border: 1px ${colors.bw.gray};
    border-radius: 6px;
    width: ${props => (props.width ? props.width : '130px')}
    height: ${props => (props.height ? props.height : '60px')}
    background-color: ${props => (props.color ? props.color : colors.primary.white)};
`;

export const ContentViewLeft = styled.View`
    ${CommonContentView};
    background-color: ${props => (props.color ? props.color : colors.primary.white)};
    border-top-left-radius: 6px;
    border-bottom-left-radius: 6px;
    border-bottom-width: 0.4;
    border-bottom-color: ${colors.bw.gray};
`;

export const ContentViewRight = styled.View`
    ${CommonContentView};
    background-color: ${props => (props.color ? props.color : colors.primary.white)};
    border-top-right-radius: 6px;
    border-bottom-right-radius: 6px;
    border-bottom-width: 0.4;
    border-bottom-color: ${colors.bw.gray};
`;

export const ContentViewDefault = styled.View`
    ${CommonContentView};
    border-bottom-width: 0.4;
    border-bottom-color: ${colors.bw.gray};
`;

export const CustomTextOperators = styled.Text`
    font-size: 20px;
    color: #777777;
`;

export const CustomTouchableHighlight = styled.TouchableHighlight`
    flex-grow: 1;
    border-radius: 6px;
    background-color: ${props => (props.color ? props.color : colors.primary.white)};
`;
