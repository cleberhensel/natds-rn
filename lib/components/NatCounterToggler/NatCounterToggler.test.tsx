import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { NatCounterToggler } from './NatCounterToggler';

describe('NatCounterToggler', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should render the component correctly', () => {
    const CounterToggler = renderer.create(<NatCounterToggler />).toJSON();

    expect(CounterToggler).toMatchSnapshot();
  });
  it('Should simulate the decrease click on the component', () => {
    const onDecreaseEvent = jest.fn();
    const wrapper = shallow(<NatCounterToggler onDecrease={onDecreaseEvent} />);

    wrapper.getElement().props.children[0].props.onPress();
    expect(onDecreaseEvent.mock.calls.length).toBe(1);
  });
  it('Should simulate the increase click on the component', () => {
    const onIncreaseEvent = jest.fn();
    const wrapper = shallow(<NatCounterToggler onIncrease={onIncreaseEvent} />);

    wrapper.getElement().props.children[2].props.onPress();
    expect(onIncreaseEvent.mock.calls.length).toBe(1);
  });
  it('Should limit the counter on the component and deny further increment upon htting the maximum limit', () => {
    let counter = 1;
    const onIncrease = () => {
      counter += 1;
    };
    const wrapper = shallow(
      <NatCounterToggler
        counter={counter}
        min={1}
        max={2}
        onIncrease={onIncrease}
      />,
    );

    for (let index = 0; index < 3; index += 1) {
      // add button
      wrapper.getElement().props.children[2].props.onPress();
    }
    // add button
    expect(wrapper.getElement().props.children[2].props.disabled).toBe(true);
  });
  it('Should limit the counter on the component and deny further decrement upon htting the minimum limit', () => {
    let counter = 2;
    const onDecrease = () => {
      counter -= 1;
      // eslint-disable-next-line no-use-before-define
      wrapper.update();
    };
    const onIncrease = () => {
      counter += 1;
      // eslint-disable-next-line no-use-before-define
      wrapper.update();
    };
    const wrapper = shallow(
      <NatCounterToggler
        counter={counter}
        min={1}
        max={4}
        onDecrease={onDecrease}
        onIncrease={onIncrease}
      />,
    );

    for (let index = 0; index < 3; index += 1) {
      // add button
      wrapper.getElement().props.children[2].props.onPress();
    }
    for (let index = 0; index < 11; index += 1) {
      // decrease button
      wrapper.getElement().props.children[0].props.onPress();
    }
    // decrease button
    expect(wrapper.getElement().props.children[0].props.disabled).toBe(true);
  });
});
