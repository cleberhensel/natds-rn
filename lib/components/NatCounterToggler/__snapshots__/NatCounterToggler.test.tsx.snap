// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NatCounterToggler Should render the component correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "stretch",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#BBBBBB",
        "borderRadius": 6,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-between",
        "width": 130,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 6,
          "flexGrow": 1,
        },
      ]
    }
  >
    <View
      color="#EEEEEE"
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#EEEEEE",
            "borderBottomColor": "#BBBBBB",
            "borderBottomLeftRadius": 6,
            "borderBottomWidth": 0.4,
            "borderTopLeftRadius": 6,
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
          null,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#777777",
              "fontSize": 20,
            },
          ]
        }
      >
        -
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#BBBBBB",
          "borderBottomWidth": 0.4,
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#777777",
            "fontSize": 20,
          },
        ]
      }
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 6,
          "flexGrow": 1,
        },
      ]
    }
  >
    <View
      color="#FFFFFF"
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderBottomColor": "#BBBBBB",
            "borderBottomRightRadius": 6,
            "borderBottomWidth": 0.4,
            "borderTopRightRadius": 6,
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
          null,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#777777",
              "fontSize": 20,
            },
          ]
        }
      >
        +
      </Text>
    </View>
  </View>
</View>
`;
